### PR TITLE
Graphql endpoint and queries for VSphere and OVirt providers

### PR DESCRIPTION
**Pull request by gildub**
Add GraphQL feature, which is activated only if environment variable `GRAPHQL_ENABLED` is enabled (for example GRAPHQL_ENABLE=true)

With following endpoinds:
- `/graphql` used by API clients 
- `/graphql/playground` built-in client (See screenshot below) used only in development. (Need to add a debug flag)

Offers queries to traverse VSphere and OVirt providers all the way down to VMs, through datacenters, clusters, hosts, networks, datastores, disks.

Screenshot from 2021-10-08 18-23-12

https://github.com/konveyor/forklift-controller/issues/359

### CODE REVIEW

File: pkg/controller/provider/web/graph/model/models_gen.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Code generated by github.com/99designs/gqlgen, DO NOT EDIT.
+
+package model
+
+type VsphereProvider struct {
+	ID      string `json:"id"`
+	Name    string `json:"name"`
+	Kind    string `json:"kind"`
+	Product string `json:"product"`
```
Comment (jortel) [inline pkg/controller/provider/web/graph/model/models_gen.go:9]: What is a product?

Comment (gildub) [reply pkg/controller/provider/web/graph/model/models_gen.go:9]: That's a reminiscent of an VMware object in Opal schema. 
Never mind, I'll remove it.

**Comment by jortel** [comment]
Why is the generator a dependency?
Seems the generator would be used initially to generate the code based on the yaml file.  And, potentially by the makefile with could get it following the existing pattern of `controller-gen`.

See: https://github.com/konveyor/forklift-controller/issues/359#issuecomment-927919363

**Comment by gildub** [comment]
@jortel, sure but where do you see it defined as a dependency? I might be missing something meanwhile I didn't add the generator to the modules.

File: go.mod [hunk 2]
```diff
@@ -3,6 +3,7 @@ module github.com/konveyor/forklift-controller
 go 1.14
 
 require (
+	github.com/99designs/gqlgen v0.14.0
```
Comment (jortel) [inline go.mod:6]: This isn't a code generator?

Comment (gildub) [reply go.mod:6]: Oh I see, that's actually both, the GraphQL library which uses a generator.

File: pkg/controller/provider/web/base/handler.go [hunk 3]
```diff
@@ -1,21 +1,23 @@
...
-	"net/http"
-	"strconv"
-	"strings"
-	"time"
 )
 
 //
 // Root - all routes.
 const (
+	GraphqlRoot   = "graphql"
```
Comment (jortel) [inline pkg/controller/provider/web/base/handler.go:20]: This should go in the GraphHander .go file.

File: pkg/controller/provider/web/graph/doc.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"github.com/konveyor/controller/pkg/inventory/container"
+	libweb "github.com/konveyor/controller/pkg/inventory/web"
+	"github.com/konveyor/forklift-controller/pkg/controller/provider/web/base"
+)
+
+//
+// Routes
+const (
+	Root = base.GraphqlRoot
```
Comment (jortel) [inline pkg/controller/provider/web/graph/doc.go:12]: Let's just define the Root = "graphql" here.

File: pkg/controller/provider/web/graph/graphql.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/99designs/gqlgen/graphql/handler"
+	"github.com/99designs/gqlgen/graphql/playground"
+)
+
+var log = logging.WithName("web|graphql")
+
+//
+// Routes.
+const (
+	GraphqlRoot = Root
```
Comment (jortel) [inline pkg/controller/provider/web/graph/graphql.go:18]: Let's just define the GraphRoot = "graphql" here.

File: pkg/controller/provider/web/graph/graphql.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	base.Handler
+}
+
+//
+// Add routes to the `gin` router.
+func (h *GraphHandler) AddRoutes(e *gin.Engine) {
+	e.POST(GraphqlRoot, h.GraphqlHandler())
+	e.GET(GraphqlRoot+"/playground", playgroundHandler())
+}
+
```
Comment (jortel) [inline pkg/controller/provider/web/graph/graphql.go:33]: To be consistent with other handlers, recommend something like:
```
e.POST(GraphqlRoot, h.Post)
e.GET(GraphqlRoot+"/playground", h.Get)

func (h GraphHandler) Post(ctx *gin.Context) {
    err := r.updateContext(ctx)
    if err != nil {
        ctx.Status(http.StatusInternalServerError)
        return
    }

    r.qqlh.ServeHTTP(ctx.Writer, ctx.Request)
}

func (h GraphHandler) Get(ctx *gin.Context) {
    err := r.updateContext(ctx)
    if err != nil {
        ctx.Status(http.StatusInternalServerError)
        return
    }

    r.playground.ServeHTTP(ctx.Writer, ctx.Request)
}

func (h *GraphHandler) updateContext(ctx *gin.Context) (err error) {
    # inject container into the ctx.Request.Context
    return
}
```

The GraphHandler.gqlh and GraphHandler.playground can be assigned when the Handler is constructed in doc.go.

Comment (gildub) [reply pkg/controller/provider/web/graph/graphql.go:33]: Makes sense.

File: pkg/controller/provider/web/graph/doc.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+//
+// Build all handlers.
+func Handlers(container *container.Container) []libweb.RequestHandler {
+	return []libweb.RequestHandler{
+		&GraphHandler{
+			Handler: base.Handler{
+				Container: container,
```
Comment (jortel) [inline pkg/controller/provider/web/graph/doc.go:21]: :+1:

File: pkg/controller/provider/web/graph/graphql.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+//
+// GraphQL Handler.
+func (h GraphHandler) Post() gin.HandlerFunc {
+	handler := handler.NewDefaultServer(generated.NewExecutableSchema(generated.Config{Resolvers: &Resolver{}}))
+
+	return func(c *gin.Context) {
+		ctx := context.WithValue(c.Request.Context(), "HandlerContainer", h.Container)
+		c.Request = c.Request.WithContext(ctx)
+		c.Next()
```
Comment (jortel) [inline pkg/controller/provider/web/graph/graphql.go:44]: I don't think you want to do c.Next() here.  I think c.Next() will advance to the next handler.

File: pkg/controller/provider/web/graph/graphql.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+//
+// Add routes to the `gin` router.
+func (h *GraphHandler) AddRoutes(e *gin.Engine) {
+	e.POST(GraphqlRoot, h.Post())
+	e.GET(GraphqlRoot+"/playground", h.Get())
+}
+
+//
+// GraphQL Handler.
+func (h GraphHandler) Post() gin.HandlerFunc {
```
[the remark this answers is not in the document]
Comment (gildub) [reply pkg/controller/provider/web/graph/graphql.go:38]: Somehow I can think with function first class citizen only.

File: go.mod (hunk 2 above)
Comment (gildub) [reply go.mod:6]: @jortel, gqlgen is used here: https://github.com/konveyor/forklift-controller/pull/358/files#diff-72eef241725fe361fe8d4e1e9bbd25b25915aa2ae9085c80175da93856bcdbb6R15-R16

**Comment by jortel** [comment]
I really appreciate the work that went into this PR. However, it adds an additional API for which there is no known need / use case.
It is already out of sync with the REST API and should it be merged will require on-going maintenance.  Closing for now.  It can be re-opened should the need arise to support a GraphQL API.

**Pull request closed without merging**